packages/contracts-bedrock/src/L1/OPContractsManager.sol

```diff
@@ -363,6 +374,40 @@ contract OPContractsManager is ISemver {
         return output;
     }
 
+    /// @notice Upgrades a set of chains to the latest implementation contracts
+    /// @param _systemConfigs Array of SystemConfig contracts, one per chain to upgrade
+    /// @param _proxyAdmins Array of ProxyAdmin contracts, one per chain to upgrade
+    /// @dev This function is intended to be called via DELEGATECALL from the Upgrade Controller Safe
+    function upgrade(ISystemConfig[] calldata _systemConfigs, IProxyAdmin[] calldata _proxyAdmins) external {
+        // TODO: emit the Upgraded event
+        // TODO: enforce delegatecalling
+        Implementations memory impls = thisOPCM.implementations();
+
+        for (uint256 i = 0; i < _systemConfigs.length; i++) {
+            ISystemConfig systemConfig = _systemConfigs[i];
+            ISystemConfig.Addresses memory opChainAddrs = ISystemConfig.Addresses({
+                l1CrossDomainMessenger: systemConfig.l1CrossDomainMessenger(),
+                l1ERC721Bridge: systemConfig.l1ERC721Bridge(),
+                l1StandardBridge: systemConfig.l1StandardBridge(),
+                disputeGameFactory: systemConfig.disputeGameFactory(),
+                optimismPortal: systemConfig.optimismPortal(),
+                optimismMintableERC20Factory: systemConfig.optimismMintableERC20Factory(),
+                gasPayingToken: address(0)
+            });
```

Comment:
Suggestion from @mds1 (pairing call): f
1. extract this into an opcm getter to reduce boiler plate
2. add a getter to the system config which returns this struct to make future upgrades cleaner and more efficient.

Reply:
Add todo comments for these items.